Logging/Loading TorchScripted model

shrinath-suresh opened the pull request:
Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

## What changes are proposed in this pull request?

Current `mlflow.pytorch` library doesnt not have provision for logging/saving the torchscripted model. The new inclusion facilitates logging/saving of the torchscripted model into mlflow.

## How is this patch tested?

Tested with MNIST handwritten digits and IrisClassification example. Added necessary unit tests for the same.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Users can now log and load torchscripted model using `mlflow.pytorch` library.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [x] `language/python`: Python APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

shrinath-suresh committed:
Saving/Logging torchscripted model. Added examples and UTs

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

harupy commented on mlflow/pytorch/__init__.py:
```diff
@@ -325,7 +325,11 @@ def save_model(
-    torch.save(pytorch_model, model_path, pickle_module=pickle_module, **kwargs)
+    if isinstance(pytorch_model, torch.jit.ScriptModule):
+        model_path = os.path.join(model_data_path, _SERIALIZED_TORCH_MODEL_FILE_NAME)
+        torch.jit.ScriptModule.save(pytorch_model, model_path)
```
Is there any difference between `torch.jit.ScriptModule.save` and `torch.jit.save`?

harupy commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,104 @@
+# pylint: disable=W0221
+# pylint: disable=W0613
+# pylint: disable=W0223
```
why ignore these pyling rules?

shrinath-suresh replied:
`W0223: Method '_forward_unimplemented' is abstract in class 'Module' but is not overridden (abstract-method)` - Eventhough the forward method is implemented in IrisClassifier class, pylint throws the error. Hence suppressed W0223. 

i guess W0221 and W0613 are not need for this example. Mostly on pytorch lightning examples, batch_idx would be added as part of training step and since its not used the pylint error is thrown. We cannot remove the batch_idx argument as it will result in non matching abstract method signature. 

But for this example W0221 and W0613  can be removed i believe, i will add a new commit for this change.

harupy commented on examples/pytorch/torchscript/IrisClassification/README.md:
```diff
@@ -0,0 +1,40 @@
+## MNIST example with MLFlow
```
```suggestion
## MNIST example with MLflow
```
nit

harupy commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,104 @@
+    test_model(scripted_model, X_test, y_test)
+
+    mlflow.tracking.set_tracking_uri(args.tracking_uri)
+    mlflow.start_run()
+    mlflow.pytorch.log_model(scripted_model, "model")  # logging scripted model
+    uri_path = mlflow.get_artifact_uri()
+    mlflow.pytorch.load_model(uri_path + "/model")  # loading scripted model
+    mlflow.end_run()
```
Could we use `with mlflow.start_run()` instead of `mlflow.start_run()` and `mlflow.end_run()`?

harupy commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,104 @@
+    return model
+
+
+def test_model(model, X_test, y_test):
+    with torch.no_grad():
+        predict_out = model.forward(X_test)
+        _, predict_y = torch.max(predict_out, 1)
+
+        print(
+            "prediction accuracy", float(accuracy_score(y_test.cpu(), predict_y.cpu())),
+        )
```
Do we need to call `model.eval` before the prediction?

harupy commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,104 @@
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
+
+    for epoch in range(epochs):
+        out = model.forward(X_train)
```
Which one is more common, `model.forward(X_train)` vs. `model(X_train)`?

harupy commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,104 @@
+    )
+
+    args = parser.parse_args()
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
```suggestion
```

because `device` is already defined at line 32.

harupy commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,104 @@
+    parser = argparse.ArgumentParser(description="Iris Classification Torchscripted model")
+
+    parser.add_argument(
+        "--tracking-uri", type=str, default="http://localhost:5000/", help="mlflow tracking uri"
```
If we set `tracking_uri` to `http://localhost:5000`, we'd need to launch an MLflow server before running this example, which I thinks is a bit cumbersome.

shrinath-suresh replied:
All the examples expects user to launch mlflow ui before running the example. this argument just facilitates user to provide a tracking uri.. For example if mlflow server is running in Amazon Ec2, user can simply specify an argument `--tracking-uri http:<EC2-URL>:5000`.

harupy commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,104 @@
+        "--tracking-uri", type=str, default="http://localhost:5000/", help="mlflow tracking uri"
+    )
+    parser.add_argument(
+        "--epochs", type=int, default=100, help="number of epochs to run (default: 50)"
```
```suggestion
        "--epochs", type=int, default=100, help="number of epochs to run (default: 100)"
```

harupy commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,104 @@
+    mlflow.tracking.set_tracking_uri(args.tracking_uri)
+    mlflow.start_run()
+    mlflow.pytorch.log_model(scripted_model, "model")  # logging scripted model
+    uri_path = mlflow.get_artifact_uri()
+    mlflow.pytorch.load_model(uri_path + "/model")  # loading scripted model
```
```suggestion
    model_path = mlflow.get_artifact_uri("model")
    mlflow.pytorch.load_model(model_path)  # loading scripted model
```

should work.

harupy replied:
Could you make predictions using the loaded model?

harupy commented on examples/pytorch/torchscript/MNIST/README.md:
```diff
@@ -0,0 +1,40 @@
+## MNIST example with MLFlow
```
```suggestion
## MNIST example with MLFlow
```

ditto

harupy commented on examples/pytorch/torchscript/MNIST/mnist_torchscript.py:
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python
+# coding: utf-8
```
```suggestion
```

harupy commented:
Thanks for the PR! Left a few comments. Other things we need to consider are:

1. Does the current `_PyTorchWrapper` implementation work when `pytorch_model` is a scripted model?

https://github.com/mlflow/mlflow/blob/5e22def3e476070012cf2e6ef7d750eb0bec0e60/mlflow/pytorch/__init__.py#L468-L494

2. How can we load back a model in pytorch < 1.6.0 where`torch.load` doesn't automatically detect a model type (eager or scripted)?

```
try:
    # load the model as an eager model.
    return torch.load(...)
except:
    # If fails, assume the model is a scripted model
    return torch.jit.load(...)
```

shrinath-suresh commented:
> Thanks for the PR! Left a few comments. Other things we need to consider are:
> 
> 1. Does the current `_PyTorchWrapper` implementation work when `pytorch_model` is a scripted model?
> 
> https://github.com/mlflow/mlflow/blob/5e22def3e476070012cf2e6ef7d750eb0bec0e60/mlflow/pytorch/__init__.py#L468-L494
> 
> 1. How can we load back a model via `mlflow.pytorch.load_model` in pytorch < 1.6.0 where`torch.load` doesn't automatically detect a model type (eager or scripted)? One apporach could be:
> 
> ```python
> if torch_version >= (1, 6, 0):
>     model = torch.load(path)
> else:
>     try:
>         # load the model as an eager model.
>         model = torch.load(...)
>     except:
>         # If fails, assume the model is a scripted model
>         model = torch.jit.load(...)
> ```

Thanks for the review and comments. I am working on addressing the review comments. I will check on the above mentioned items and get back to you.

shrinath-suresh committed:
Addressing review comments

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh committed:
Using _PyTorchWrapper for prediction

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh commented:
> Thanks for the PR! Left a few comments. Other things we need to consider are:
> 
> 1. Does the current `_PyTorchWrapper` implementation work when `pytorch_model` is a scripted model?
> 
> https://github.com/mlflow/mlflow/blob/5e22def3e476070012cf2e6ef7d750eb0bec0e60/mlflow/pytorch/__init__.py#L468-L494
> 
> 1. How can we load back a model via `mlflow.pytorch.load_model` in pytorch < 1.6.0 where`torch.load` doesn't automatically detect a model type (eager or scripted)? One apporach could be:
> 
> ```python
> if torch_version >= (1, 6, 0):
>     model = torch.load(path)
> else:
>     try:
>         # load the model as an eager model.
>         model = torch.load(...)
>     except:
>         # If fails, assume the model is a scripted model
>         model = torch.jit.load(...)
> ```

@harupy  I updated the [IrisClassification](https://github.com/mlflow/mlflow/pull/3557/commits/fa1676e3156b8e5adfb957e00230691b5cb179e5) example to use `_PyTorchWrapper` class for prediction. It is working as expected. I am working on the next item now.

harupy commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,113 @@
+        with torch.no_grad():
+            test_datapoint = pd.DataFrame([[4.4000, 3.0000, 1.3000, 0.2000]])
+            actual = "setosa"
+            pywrap = _PyTorchWrapper(loaded_pytorch_model)
```
`_PyTorchWrapper` is not meant to be used like this. You can use `mlfow.pyfunc.load_model` instead which returns a model wrapped by `_PyTorchWrapper`.

shrinath-suresh replied:
Okie. Loading the scripted model and making prediction is already part of Unit Test - https://github.com/mlflow/mlflow/blob/fa1676e3156b8e5adfb957e00230691b5cb179e5/tests/pytorch/test_pytorch_model_export.py#L307 . Shall i revert to the old code where we make the prediction or you want me to update the code and make predictions based on pyfunc model ?

harupy replied:
I was expecting something like:

```python
loaded_pytorch_model = mlflow.pytorch.load_model(model_path)
preds = loaded_pytorch_model(X)

print(preds)
```

shrinath-suresh replied:
Thats how i had it earlier - https://github.com/mlflow/mlflow/pull/3557/commits/62690812d7dca3b7566936681373a3cc7a82b281 . will revert this commit.

harupy commented on examples/pytorch/torchscript/MNIST/mnist_torchscript.py:
```diff
@@ -0,0 +1,194 @@
+# pylint: disable=W0223
+
+from __future__ import print_function
```
```suggestion
from __future__ import print_function
```

because we no longer support python 2

shrinath-suresh commented:
> > Thanks for the PR! Left a few comments. Other things we need to consider are:
> > 
> > 1. Does the current `_PyTorchWrapper` implementation work when `pytorch_model` is a scripted model?
> > 
> > https://github.com/mlflow/mlflow/blob/5e22def3e476070012cf2e6ef7d750eb0bec0e60/mlflow/pytorch/__init__.py#L468-L494
> > 
> > 
> > 1. How can we load back a model via `mlflow.pytorch.load_model` in pytorch < 1.6.0 where`torch.load` doesn't automatically detect a model type (eager or scripted)? One apporach could be:
> > 
> > ```python
> > if torch_version >= (1, 6, 0):
> >     model = torch.load(path)
> > else:
> >     try:
> >         # load the model as an eager model.
> >         model = torch.load(...)
> >     except:
> >         # If fails, assume the model is a scripted model
> >         model = torch.jit.load(...)
> > ```
> 
> @harupy I updated the [IrisClassification](https://github.com/mlflow/mlflow/pull/3557/commits/fa1676e3156b8e5adfb957e00230691b5cb179e5) example to use `_PyTorchWrapper` class for prediction. It is working as expected. I am working on the next item now.

Tested the torchscripted IrisExample in all 3 versions. In torch 1.5 and 1.6 am able to load the model directly using `torch.load`. Only in 1.4, we need `torch.jit.load()` . As suggested, we can add a version check for this.

shrinath-suresh committed:
Making prediction using model(input)

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh committed:
Removing print function and pylint disable checks

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh committed:
Adding specific condition for torch=1.4.0 to use torch.jit.load

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh commented:
> > > Thanks for the PR! Left a few comments. Other things we need to consider are:
> > > 
> > > 1. Does the current `_PyTorchWrapper` implementation work when `pytorch_model` is a scripted model?
> > > 
> > > https://github.com/mlflow/mlflow/blob/5e22def3e476070012cf2e6ef7d750eb0bec0e60/mlflow/pytorch/__init__.py#L468-L494
> > > 
> > > 
> > > 1. How can we load back a model via `mlflow.pytorch.load_model` in pytorch < 1.6.0 where`torch.load` doesn't automatically detect a model type (eager or scripted)? One apporach could be:
> > > 
> > > ```python
> > > if torch_version >= (1, 6, 0):
> > >     model = torch.load(path)
> > > else:
> > >     try:
> > >         # load the model as an eager model.
> > >         model = torch.load(...)
> > >     except:
> > >         # If fails, assume the model is a scripted model
> > >         model = torch.jit.load(...)
> > > ```
> > 
> > 
> > @harupy I updated the [IrisClassification](https://github.com/mlflow/mlflow/pull/3557/commits/fa1676e3156b8e5adfb957e00230691b5cb179e5) example to use `_PyTorchWrapper` class for prediction. It is working as expected. I am working on the next item now.
> 
> Tested the torchscripted IrisExample in all 3 versions. In torch 1.5 and 1.6 am able to load the model directly using `torch.load`. Only in 1.4, we need `torch.jit.load` . As suggested, we can add a version check for this.

@harupy Since, we need to perform `>=` comparison for version, i thought of using `packaging.version` utility. Looks like `packaging` pip package is not used in mlflow repo. Is it okay to add it in `small-requirements.txt` or we should perform string parse and do the comparison. Your thoughts ?

harupy commented:
I think we can use `distutils.version.LooseVersion`.

harupy commented:
Does the current `_PytorchWrapper` implementation works for both eager and scripted model?

shrinath-suresh commented:
> Does the current `_PytorchWrapper` implementation work for both eager and scripted model?

Yes , the unit test covers both eager and scripted implementation. Anyways , i will cross check once again and update it here.

harupy commented on examples/pytorch/torchscript/IrisClassification/conda.yaml:
```diff
@@ -0,0 +1,14 @@
+  - sklearn
+  - cloudpickle==1.6.0
+  - boto3
+name: mlflow-env
```
Can you remove the `name` field here? I think it's unncessary.

harupy commented on mlflow/pytorch/__init__.py:
```diff
@@ -15,6 +15,7 @@
 import cloudpickle
 import numpy as np
 import pandas as pd
+from packaging.version import Version
```
Can you remove packaging here as we discussed?

shrinath-suresh replied:
yeah sure  i will replace it with LooseVersion and upload the latest code today.

shrinath-suresh replied:
Updated the code.

harupy commented on mlflow/pytorch/__init__.py:
```diff
@@ -363,11 +368,17 @@ def _load_model(path, **kwargs):
         # information about the pickle module that should be used by PyTorch to load it
         model_path = os.path.join(path, "model.pth")
+
+        local_path = os.path.dirname(path)
+        pytorch_conf = _get_flavor_configuration(model_path=local_path, flavor_name=FLAVOR_NAME)
+        pytorch_version = pytorch_conf["pytorch_version"]
```
Why do we need to use the pytorch version used to saves the model?

shrinath-suresh replied:
https://github.com/chauhang/mlflow/blob/1eec412236ae8f06c2dabd2dde28175ed6bb06a1/tests/pytorch/test_pytorch_model_export.py#L438 - This test specifically alters the pytorch version and without this condition, pytorch version would be set to 1.6 and  loading the model using `torch.load`  results in exception.

harupy replied:
I think this issue can be fixed like this:
: https://github.com/harupy/mlflow/commit/cba164e1989e7857170a26752538a31bd94fd7c5

shrinath-suresh replied:
Thanks . Let me check.

harupy commented on mlflow/pytorch/__init__.py:
```diff
@@ -395,7 +406,16 @@ def _load_model(path, **kwargs):
+        try:
+            # load the model as an eager model.
+            return torch.load(model_path, **kwargs)
+        except Exception: # pylint: disable=W0703
```
Can you replace `W0703` to `catching-too-general-exception`?

shrinath-suresh replied:
sure

shrinath-suresh committed:
Using LooseVersion to compare pytorch version number

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh committed:
Fixing Unit Tests

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

harupy commented on mlflow/pytorch/__init__.py:
```diff
@@ -363,6 +368,7 @@ def _load_model(path, **kwargs):
     :param kwargs: Additional kwargs to pass to the PyTorch ``torch.load`` function.
     """
     import torch
+    pytorch_version = torch.__version__
```
Do we need to define this variable? Can we directly pass `torch.__version__` to `LooseVersion`

shrinath-suresh replied:
Done

shrinath-suresh committed:
Removing pytorch_version variable and direction using torch.__version__

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh committed:
Applying black

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

harupy commented on examples/pytorch/torchscript/IrisClassification/conda.yaml:
```diff
@@ -0,0 +1,13 @@
+- pip:
+  - sklearn
+  - cloudpickle==1.6.0
+  - boto3
```
nit: linebreak

harupy commented on mlflow/pytorch/__init__.py:
```diff
@@ -325,7 +326,11 @@ def save_model(
         f.write(pickle_module.__name__)
     # Save pytorch model
     model_path = os.path.join(model_data_path, _SERIALIZED_TORCH_MODEL_FILE_NAME)
-    torch.save(pytorch_model, model_path, pickle_module=pickle_module, **kwargs)
+    if isinstance(pytorch_model, torch.jit.ScriptModule):
+        model_path = os.path.join(model_data_path, _SERIALIZED_TORCH_MODEL_FILE_NAME)
```
Can we remove `model_path` at line 330 because it's the same as the one at line 328?

shrinath-suresh committed:
Removing duplicated model_path variable

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh committed:
Adding linebreak

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

smurching commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,107 @@
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Iris Classification Torchscripted model")
+
+    parser.add_argument(
```
Similar to the discussion in https://github.com/mlflow/mlflow/pull/3601#discussion_r513088940, in order to follow MLflow best practices could we remove tracking URI as an argument to the script & also remove it from the MLproject file?

shrinath-suresh replied:
removed tracking uri argument and updated readme files with necessary details

smurching commented on examples/pytorch/torchscript/MNIST/MLproject:
```diff
@@ -0,0 +1,20 @@
+      epochs: {type: int, default: 5}
+      batch_size: {type: int, default: 64}
+      learning_rate: {type: float, default: 1e-3}
+      tracking_uri: {type: str, default: 'http://localhost:5000'}
```
```suggestion
```

shrinath-suresh replied:
removed tracking uri argument and updated readme files with necessary details

smurching commented on examples/pytorch/torchscript/IrisClassification/MLproject:
```diff
@@ -0,0 +1,15 @@
+  main:
+    parameters:
+      epochs: {type: int, default: 50}
+      tracking_uri: {type: str, default: 'http://localhost:5000'}
```
```suggestion
```

shrinath-suresh replied:
removed tracking uri argument and updated readme files with necessary details

smurching commented on examples/pytorch/torchscript/IrisClassification/iris_classification.py:
```diff
@@ -0,0 +1,107 @@
+
+    mlflow.tracking.set_tracking_uri(args.tracking_uri)
+    with mlflow.start_run() as run:
+        mlflow.pytorch.log_model(scripted_model, "model")  # logging scripted model
```
```suggestion
        mlflow.pytorch.log_model(scripted_model)  # logging scripted model
```

Nit: I think the default model path is "model", so no need to specify the model path here :)

harupy replied:
@smurching @shrinath-suresh 

`artifact_path` is not an optional argument.

https://github.com/mlflow/mlflow/blob/master/mlflow/pytorch/__init__.py#L69

shrinath-suresh replied:
@harupy Thanks harupy. You are right. its not an optional argument.

smurching commented on mlflow/pytorch/__init__.py:
```diff
@@ -325,7 +326,10 @@ def save_model(
     # Save pytorch model
     model_path = os.path.join(model_data_path, _SERIALIZED_TORCH_MODEL_FILE_NAME)
-    torch.save(pytorch_model, model_path, pickle_module=pickle_module, **kwargs)
+    if isinstance(pytorch_model, torch.jit.ScriptModule):
```
Could we update the docstrings of `log_model` and `load_model` to mention that you can now also log a TorchScript model (i.e. the output of `torch.jit.trace` or `torch.jit.script`)?

smurching replied:
It'd also be good to update the documentation of ``mlflow.pytorch.load_model`` to clarify that we load back either an eager model (subclass of `torch.nn.Module`) or a TorchScripted model, depending on whether an eager or TorchScripted model was originally logged.

shrinath-suresh replied:
Updated the doc strings by mentioning the support of scripted model. In the example section, added the code to log and save model as well

shrinath-suresh replied:
We will raise a separate PR for updating documentation of `mlflow.pytorch`

smurching reviewed (commented):
Small comments otherwise looks good!

harupy commented:
@shrinath-suresh Looks like the conlicts block GitHub Actions.

shrinath-suresh committed:
Removing tracking uri parser argument and updated readme files

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh committed:
Updating doc strings of mlflow.pytorch log and save model

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh committed:
Fetching latest changes and resolving conflicts

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

smurching committed:
Update docstring

Signed-off-by: Sid Murching <sid.murching@databricks.com>

smurching reviewed (approved):
@shrinath-suresh would you be able to reintroduce the `scripted_model` fixture to fix the tests?

I've pushed also pushed a small edit to your docstring - LGTM once the tests are fixed!

harupy commented on mlflow/pytorch/__init__.py:
```diff
@@ -391,6 +392,35 @@ def save_model(
+                tmp_extra_files_dir.path(), posixpath.join(path, saved_extra_files_dir_subpath)
+            )
+
+    torch.save(pytorch_model, model_path, pickle_module=pickle_module, **kwargs)
```
test

harupy replied:
@shrinath-suresh Can you remove `torch.save` here?

shrinath-suresh committed:
Removing duplicate torch.save and adding scripted_model condition for save artifacts UTs

Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

shrinath-suresh commented:
@smurching Thank you very much. I have fixed the tests.

smurching committed:
Fixed Iris README

Updated the Iris example README to refer to the Iris dataset, rather than MNIST

smurching committed:
Fix MNIST README instructions for running project

smurching committed:
Merge branch 'master' into pytorch-torchscript-save

Signed-off-by: Sid Murching <sid.murching@databricks.com>
